tc/tc_util.c

```diff
@@ -190,39 +190,42 @@ static const struct rate_suffix {
 	{ NULL }
 };
 
-int parse_percent_rate(char *rate, const char *str, char *dev)
+int parse_percent_rate(char *rate, const char *str, const char *dev)
 {
-	long max_rate_bits;
+	long mbit;
 	int ret, saved_errno;
 	double perc, rate_bits;
-	char *str_perc, *p;
+	char *str_perc;
 
 	if (!dev[0]) {
 		fprintf(stderr, "No device specified; specify device to rate limit by percentage\n");
 		return -1;
 	}
 
-	if (read_prop(dev, "speed", &max_rate_bits))
+	if (read_prop(dev, "speed", &mbit) == 1) {
+		fprintf(stderr, "Link speed for %s is currently not known, rate limits for %s currently cannot be specified in %%!\n", dev, dev);
+		return -1;
+	}
+
+	if (read_prop(dev, "speed", &mbit))
 		return -1;
 
 	ret = sscanf(str, "%m[0-9.%]", &str_perc);
 	if (ret != 1)
 		goto malf;
 
-	/* Make sure there's only one percent sign and it's at the end */
-	perc = strtod(str_perc, &p);
-	if (*p != '%' || *(p++) != '\0')
+	if (parse_percent(&perc, str_perc))
 		goto malf;
 
 	saved_errno = errno;
```

Comment:
Since `parse_percent()` figures out whether is overflow and returns 1 in that case, we don't need to try to use `errno` to figure this out. You can remove the declaration of `saved_errno`, this line, and the check below of whether `saved_errno == ERANGE`